From d330e08ca4cc32115778dd646615e5f34a0df644 Mon Sep 17 00:00:00 2001
From: "Brett T. Warden" <brett.t.warden@intel.com>
Date: Wed, 8 Jan 2025 14:57:02 -0800
Subject: [PATCH] Add fallback to old cgroup slice directory (#7)

* Add fallback to old cgroup slice directory

Handle the differences between cgroupv1 to cgroupv2 automatically.
---
 main.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/main.c b/main.c
index c22c134006ba..d3be30512ca1 100644
--- a/main.c
+++ b/main.c
@@ -33,6 +33,7 @@
 #include <unistd.h>
 #include <time.h>
 
+#define SLICE_DIR_OLD "/sys/fs/cgroup/systemd/system.slice"
 #define SLICE_DIR "/sys/fs/cgroup/system.slice"
 #define SYS_DIR "/usr/share/clr-service-restart"
 #define USER_DIR "/etc/clr-service-restart"
@@ -153,6 +154,7 @@ int main(int argc, char **argv)
 {
 	bool noop = false;
 	bool all = false;
+	char *slice_dir = SLICE_DIR;
 
 	/* parse args */
 	if (argc > 1) {
@@ -247,9 +249,13 @@ int main(int argc, char **argv)
 	}
 
 	/* do restarts */
-	DIR *d = opendir(SLICE_DIR);
+	DIR *d = opendir(slice_dir);
+	if (!d && (d = opendir(SLICE_DIR_OLD))) {
+		/* fallback to old location */
+		slice_dir = SLICE_DIR_OLD;
+	}
 	if (!d) {
-		perror("opendir()");
+		fprintf(stderr, "Unable to identify processes to update\n");
 		exit(EXIT_FAILURE);
 	}
 	/* Loop over all the units in the system slice */
@@ -313,7 +319,7 @@ int main(int argc, char **argv)
 
 nofilter:
 		/* open the tasks file */
-		if (asprintf(&n, SLICE_DIR "/%s/tasks", e->d_name) < 1) {
+		if (asprintf(&n, "%s/%s/tasks", slice_dir, e->d_name) < 1) {
 			perror("asprintf()");
 			exit(EXIT_FAILURE);
 		}
-- 
2.47.1

